third_party/move/move-vm/runtime/src/storage/implementations/unsync_module_storage.rs

```diff
@@ -0,0 +1,623 @@
+            Verified(module) => return Ok(module.clone()),
+            Deserialized(compiled_module, size)
+                if compiled_module.num_immediate_dependencies() == 0 =>
+            {
+                let module = Arc::new(
+                    self.runtime_environment.build_verified_module(
+                        self.runtime_environment
+                            .build_partially_verified_module(compiled_module.clone(), *size)?,
+                        &[],
+                    )?,
+                );
+                *entry = Verified(module.clone());
+                return Ok(module);
+            },
+            _ => {
+                // Otherwise, do drop the borrow, so we can traverse the transitive closure.
+                drop(module_storage);
+            },
+        }
+
+        self.visit_and_verify_all_transitive_dependencies(address, module_name, visited)
+    }
+
+    /// Visits the dependencies of the given module. If dependencies form a cycle (which
+    /// should not be the case as we check this when modules are added to the module
+    /// storage), an error is returned.
+    ///
+    /// Important: this implementation **does not** load transitive friends. While it is
+    /// possible to view friends as `used-by` relation, it cannot be checked fully. For
+    /// example, consider the case when we have four modules A, B, C, D and let `X --> Y`
+    /// be a dependency relation (Y is a dependency of X) and `X ==> Y ` a friend relation
+    /// (X declares Y a friend). Then consider the case `A --> B <== C --> D <== A`. Here,
+    /// if we opt for `used-by` semantics, there is a cycle. But it cannot be checked,
+    /// since, A only sees B and D, and C sees B and D, but both B and D do not see any
+    /// dependencies or friends. Hence, A cannot discover C and vice-versa, making detection
+    /// of such corner cases only possible if **all existing modules are checked**, which
+    /// is clearly infeasible.
+    fn visit_and_verify_all_transitive_dependencies(
+        &self,
+        address: &AccountAddress,
+        module_name: &IdentStr,
+        visited: &mut BTreeSet<ModuleId>,
+    ) -> PartialVMResult<Arc<Module>> {
+        use ModuleStorageEntry::*;
+
+        // Step 1: verify compiled module locally.
+        let compiled_module = self.fetch_deserialized_module(address, module_name)?;
+        let size = self.fetch_module_size_in_bytes(address, module_name)?;
+        let partially_verified_module = self
+            .runtime_environment
+            .build_partially_verified_module(compiled_module, size)?;
+
+        // Step 2: visit all dependencies and collect them for later verification.
+        let mut verified_immediate_dependencies = vec![];
+        for (addr, name) in partially_verified_module.immediate_dependencies_iter() {
+            let module_id = ModuleId::new(*addr, name.to_owned());
+            if !visited.insert(module_id) && self.is_deserialized(addr, name)? {
```

Comment:
why does is_deserialized correspond to a cycle? is it only possible during the traversal itself, might be a dangerous invariant if for some reason we cache smt in deserialized format?

Reply:
You can see these visited/deserialized as colours when doing DFS. If a module has been visited and verified, it is ok since we can verify A -> B and then verify C -> B.
If it is visited and not verified yet, we must be in the same traversal, since the invariant is:
1. traverse outgoing edges
2. mark verified

Reply:
If we cache it is fine, as long as it is not visited?

Reply:
I read || instead of &&, all good, makes sense.